core/controllers/acl_decorators.py

```diff
@@ -2367,16 +2426,65 @@ def test_can_access(self, story_id, *args, **kwargs):
             PageNotFoundException: The given page cannot be found.
         """
         story = story_fetchers.get_story_by_id(story_id, strict=False)
+        lowercase_abbrev_topic_name = (
+            python_utils.url_unquote_plus(abbrev_topic_name).replace('-', ' '))
+        classroom_name = kwargs.pop('classroom_name', None)
 
         if story is None:
-            raise self.PageNotFoundException
+            handle_redirection(
+                self,
+                '/learn/%s/%s/story' % (classroom_name, abbrev_topic_name),
+                self.GET_HANDLER_ERROR_RETURN_TYPE)
+            return
 
         story_is_published = False
         topic_is_published = False
         topic_id = story.corresponding_topic_id
         user_actions_info = user_services.UserActionsInfo(self.user_id)
         if topic_id:
             topic = topic_fetchers.get_topic_by_id(topic_id)
+            if topic.abbreviated_name.lower() != lowercase_abbrev_topic_name:
+                handle_redirection(
+                    self,
+                    '/learn/%s/%s/story/%s' % (
+                        classroom_name,
+                        topic.abbreviated_name.lower().replace(' ', '-'),
+                        story_id),
+                    self.GET_HANDLER_ERROR_RETURN_TYPE)
+                return
+            if classroom_name not in [None, constants.CLASSROOM_STATUS_STAGING]:
```

Comment:
There seems to be a great deal of repetition in the logic, here and elsewhere. You might want to have a classroom_services.py or similar to handle some of the common logic.